src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfoCacheService.cs

```diff
@@ -59,138 +91,124 @@ public SymbolTreeInfoCacheService()
             return _peReferenceToInfo.GetOrAdd(reference, new MetadataInfo(info, referencingProjects)).SymbolTreeInfo;
         }
 
-        public async Task<SymbolTreeInfo?> TryGetPotentiallyStaleSourceSymbolTreeInfoAsync(
+        public async ValueTask<SymbolTreeInfo?> TryGetPotentiallyStaleSourceSymbolTreeInfoAsync(
             Project project, CancellationToken cancellationToken)
         {
+            if (!project.SupportsCompilation)
+                return null;
+
+            // Kick off the work to update the data we have for this project.
+            _workQueue.AddWork(project.Id);
+
             // See if the last value produced exactly matches what the caller is asking for.  If so, return that.
             if (_projectIdToInfo.TryGetValue(project.Id, out var projectInfo))
-                return projectInfo;
+                return projectInfo.info;
 
             // If we didn't have it in our cache, see if we can load some version of it from disk.
             var info = await SymbolTreeInfo.LoadAnyInfoForSourceAssemblyAsync(project, cancellationToken).ConfigureAwait(false);
             if (info is null)
                 return null;
 
-            // attempt to add this item to the map.  But defer to whatever is in the map now if something else beat us to this.
-            return _projectIdToInfo.GetOrAdd(project.Id, info);
+            // attempt to add this item to the map.  But defer to whatever is in the map now if something else beat
+            // us to this.  Don't provide a version here so that the next time we update this data it will get
+            // overwritten with the latest computed data.
+            return _projectIdToInfo.GetOrAdd(project.Id, (semanticVersion: default, info)).info;
```

Comment:
Do we not give the semantic version of the request project here because the data from disk may be stale?

Reply:
1. that's correct.

2. so, semantic versions are 'interesting' and are not safe to use across persistence.  Consider the following:

File A_1 - Semantic Version 1.

File A gets changed in a way thtat doesn't impact semantics (e.g. a method body edit).  So you have:

File A_2 - Semantic Version 1.

There's no way to know this post VS close/open.  The 'Semantic Version' will be the same File-Version as File A_1, and that has been completely lost and cannot be recomputed.  So it's really not a safe persistence concept.